memory/ndctl_selftest.py

```diff
@@ -0,0 +1,121 @@
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2019 IBM
+# Author: Harish <harish@linux.vnet.ibm.com>
+#
+#
+
+
+import os
+import json
+
+from avocado import Test
+from avocado import main
+from avocado.utils import process, build, distro, git
+from avocado.utils.software_manager import SoftwareManager
+
+
+class NdctlTest(Test):
+    """
+    Ndctl is a user space tool to manage persistent memory devices
+    This test uses the selftests of the userspace tool to sanity check the
+    persistent memory
+    """
+
+    def get_bus_ids(self):
+        """
+        Get the provider IDs for each pmem device
+        """
+        bus_ids = []
+        os.chdir(self.sourcedir)
+        json_op = json.loads(process.system_output(
+            './ndctl/ndctl list -B', shell=True))
+        for nid in json_op:
+            for key, value in nid.items():
+                if key == 'provider':
+                    bus_ids.append(value)
+        return bus_ids
+
+    def setUp(self):
+        """
+        Prequisite for ndctl selftest on non-NFIT devices
+        """
+        smg = SoftwareManager()
+        self.url = self.params.get(
+            'url', default="https://github.com/pmem/ndctl.git")
+        self.branch = self.params.get('branch', default='master')
+
+        deps = ['gcc', 'make', 'automake', 'autoconf', 'patch']
+        detected_distro = distro.detect()
+
+        if 'SuSE' in detected_distro.name:
+            deps.extend(['ruby2.5-rubygem-asciidoctor', 'libtool',
+                         'libkmod-devel', 'libudev-devel', 'keyutils-devel',
+                         'libuuid-devel-static', 'libjson-c-devel',
+                         'systemd-devel', 'kmod-bash-completion'])
+        else:
+            # TODO: Add RHEL when support arrives
+            self.cancel('Unsupported OS %s' % detected_distro.name)
+
+        for package in deps:
+            if not smg.check_installed(package) and not smg.install(package):
+                self.cancel(
+                    "Fail to install %s required for this test." % (package))
+
+        git.get_repo(self.url, branch=self.branch,
+                     destination_dir=self.teststmpdir)
+
+        self.sourcedir = self.teststmpdir
+        os.chdir(self.sourcedir)
+
+        process.run('./autogen.sh', sudo=True, shell=True)
+        process.run(
+            "./configure CFLAGS='-g -O2' --prefix=/usr "
+            "--sysconfdir=/etc --libdir=/usr/lib64 "
+            "--enable-destructive", shell=True, sudo=True)
+
+        # TODO: remove patches once merged upstream
+        process.run('patch -p1 < %s' %
+                    self.get_data('upstream.patch'), shell=True)
```

Comment:
can we get this path hosted somewhere as it sent upstream may be we use that url to get that

Reply:
makes sense. Will adopt that change for the first patch, Second patch is specific to PPC, so would like to keep it here. I will handle that as before applying as well. Thanks.